### PR TITLE
feat(crowdstrike-recon): add new external-import connector (#5412)

### LINKED ISSUE #5412: Create a Crowdstrike Recon Connector
## Use case

I'd like to bring in data data from the Crowdstrike Recon endpoint as Incidents within OpenCTI

## Current Workaround

There currently is no workaround.

### PR DESCRIPTION
**Pull request by romain-filigran**
### Proposed changes

New `external-import` connector for the CrowdStrike Falcon Recon (Exposure Monitoring) service. It fetches Recon notifications (typosquatting domains, exposed credentials, dark web posts/replies, file leaks), converts them into OpenCTI Incidents with related observables (Domain-Name, Email-Address, User-Account, URL, Malware) plus a Markdown report attachment, supports FQL filtering on topic / type / priority, and persists `last_alert_date` between runs. Built on the shared `connectors-sdk` settings layer.

### Related issues

Closes #5412

### Maintainer review update (@SamuelHassine)

Reviewed the full connector and addressed the Copilot review pass (all threads resolved):

- converter_to_stix: guard `highlights[]`, `notification` and `breach_summary` against missing keys; default `rule_name` / `title`; base64-encode the alert attachment as a string (not bytes) so `x_opencti_files.data` is valid; derive `incident_type` from `item_type`; skip missing typosquatting domain values; add a dedicated `generate_file_content`.
- api_client: drop the misleading `HttpUrl` hint and the stored client secret; escape single quotes in FQL values; raise on non-2xx falconpy responses; guard pagination (short page + max pages); batch notification-detail lookups (100/request) to remove the N+1.
- connector: fix the set-literal log meta; track the maximum alert date; open the work as `is_multipart=True` and always close it via `to_processed(..., in_error=...)` in a `finally` block.
- docs/build: align the README support version (>= 6.9.5) and incident-type description; fix the docker-compose default comments; slim the Dockerfile build dependencies; remove the empty `utils.py`; bump `pycti` to `7.260609.0` to match `connectors-sdk` on master.
- tests: added converter, API client, settings and connector test suites; `codecov/patch` and `codecov/project` are green.
- Merged `master` to de-stale the branch (it was ~1 month behind), which also aligns it with the current `connectors-sdk` and CI.

### Status

- All CI checks green (Test external-import/crowdstrike-recon, codecov/patch, codecov/project, Build and Commit Manifest, linters, signed commits, PR title, issue link).
- 0 unresolved review threads.

### Checklist

- [x] I consider the submitted work as finished
- [x] I have signed my commits using GPG key.
- [x] I tested the code for its functionality using different use cases
- [x] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

### CODE REVIEW

File: external-import/crowdstrike-recon/src/connector/converter_to_stix.py [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+                    user_account = stix2.UserAccount(
+                        account_login=item.get("email"),
+                        allow_custom=True,
+                        object_marking_refs=[self.tlp_marking.id],
+                        custom_properties={
+                            "x_opencti_created_by_ref": self.author.id,
+                        }
+                    )
+                    inc_sco_sdo_refs.append(user_account.id)
+                    stix_objects.append(user_account)
```
[the remark this answers is not in the document]
Comment (romain-filigran) [reply external-import/crowdstrike-recon/src/connector/converter_to_stix.py:388]: @copilot apply changes based on [this feedback](https://github.com/OpenCTI-Platform/connectors/pull/6409#discussion_r3243283527)

File: external-import/crowdstrike-recon/src/connector/converter_to_stix.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+            alert_detail = self.generate_reply_content(notification_detail)
+
+        elif notification.get("item_type") == "post":
+            raw_title = notification.get("highlights")[0]
+            title = raw_title[:50] + "..." if len(raw_title) > 50 else raw_title
+            inc_description = self.generate_common_description(notification)
+            alert_detail = self.generate_post_content(notification_detail)
+
+        elif notification.get("item_type") == "file":
+            raw_title = notification.get("highlights")[0]
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/connector/converter_to_stix.py]: Fixed. Title extraction is now guarded by the `_extract_highlight_title` helper, which returns "" when `highlights` is missing, None or an empty list instead of indexing `[0]`.

File: external-import/crowdstrike-recon/src/connector/converter_to_stix.py [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+            title = raw_title[:50] + "..." if len(raw_title) > 50 else raw_title
+            inc_description = self.generate_common_description(notification)
+            alert_detail = self.generate_post_content(notification_detail)
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/connector/converter_to_stix.py]: Fixed. Added a dedicated `generate_file_content` method (documented to reuse the post `details` formatting that file notifications share) and `create_incident` now calls it for `item_type == "file"`.

File: external-import/crowdstrike-recon/src/connector/converter_to_stix.py [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+        else:
+            self.helper.connector_logger.error(
+                "Unsupported notification type, skipping alert",
+                {"item_type": notification.get("item_type"), "notification_id": notification.get("id")},
+            )
+            return []
+
+        # Create the incident
+        incident_name = notification.get("rule_name") +" : "+title
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/connector/converter_to_stix.py]: Fixed. `incident_name` is now built with defaults: `f"{rule_name or 'CrowdStrike Recon'} : {title or '--'}"`, so a missing `rule_name` or `title` no longer raises a TypeError.

File: external-import/crowdstrike-recon/src/connector/converter_to_stix.py [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+        alert_detail_bytes = alert_detail.encode("utf-8")
+        base64_bytes = base64.b64encode(alert_detail_bytes)
+        attachment_files.append(
+            {
+                "name": "alert.md",
+                "data": base64_bytes,
+                "mime_type": "text/markdown",
+                "no_trigger_import": False,
+            }
+        )
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/connector/converter_to_stix.py]: Fixed. The attachment is now `base64.b64encode(alert_detail_bytes).decode("utf-8")`, so `x_opencti_files.data` is a base64 string and the STIX object serializes correctly. Covered by `test_create_incident_attachment_is_base64_string`.

File: external-import/crowdstrike-recon/src/connector/converter_to_stix.py [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+                lines.append(f"- **{label}**: {value}\n")
+            lines.append("\n")
+
+        return "".join(lines)
+
+    def generate_exposed_data_content(self, notification_detail: dict) -> str:
+        """
+        Generate a Markdown description for exposed data notifications.
+        """
+        notification = notification_detail.get("notification")
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/connector/converter_to_stix.py]: Fixed. `generate_exposed_data_content` now uses `notification_detail.get("notification") or {}` and guards `breach_details` / `breach_summary` with `or {}`, so a missing `notification` no longer raises AttributeError.

File: external-import/crowdstrike-recon/src/connector/connector.py [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+                stix_objects_bundle = self.helper.stix2_create_bundle(stix_objects)
+                bundles_sent = self.helper.send_stix2_bundle(
+                    stix_objects_bundle,
+                    work_id=work_id,
+                    cleanup_inconsistent_bundle=True,
+                )
+
+                self.helper.connector_logger.info(
+                    "Sending STIX objects to OpenCTI...",
+                    {"bundles_sent": {str(len(bundles_sent))}},
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/connector/connector.py]: Fixed. Changed the log meta to `{"bundles_sent": str(len(bundles_sent))}` (removed the inner braces that made it a set).

File: external-import/crowdstrike-recon/src/connector/connector.py [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+            notification_detail = self.client.get_notification_detail(
+                notification_id=notification_id
+            )
+            most_recent_incident_date = notification_detail.get("notification", {}).get("created_date")
+
+            # convert notification into an OpenCTI Incident
+            stix_entities = self.converter_to_stix.create_incident(
+                notification_detail=notification_detail
+            )
+            stix_objects.extend(stix_entities)
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/connector/connector.py]: Fixed. `_collect_intelligence` now tracks the maximum `created_date` explicitly instead of taking the value from the last iteration, and the connector only persists state after a successful run, so the saved cursor cannot regress. Covered by `test_collect_intelligence_tracks_max_date_and_appends_author_tlp`.

File: external-import/crowdstrike-recon/src/connector/connector.py [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+            self.helper.connector_logger.info(message)
+
+        except (KeyboardInterrupt, SystemExit):
+            self.helper.connector_logger.info(
+                "[CONNECTOR] Connector stopped...",
+                {"connector_name": self.helper.connect_name},
+            )
+            sys.exit(0)
+        except Exception as err:
+            self.helper.connector_logger.error(str(err))
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/connector/connector.py]: Fixed. `process_message` now opens the work as `is_multipart=True` and always closes it via `to_processed(work_id, message, in_error=...)` in a `finally` block, so the work is closed on success, on error and on interrupt instead of being left in-progress. Covered by the failure/interrupt tests in `test_connector.py`.

File: external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        return notification_ids
+
+
+    def get_notification_detail(self, notification_id) -> dict:
+        """
+        :param notification_id:
+        :return:
+        """
+        notification = self.cs.get_notifications_detailed(ids=notification_id)
+        return notification.get("body").get("resources")[0]
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py]: Fixed. Added `get_notifications_details`, which batches IDs (100 per request) via `get_notifications_detailed(ids=chunk)`, and the connector now performs a single batched lookup instead of one HTTP request per notification. Covered by `test_get_notifications_details_batches_requests`.

[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py]: Fixed. Added `_raise_for_status`, which raises on a non-2xx falconpy response (surfacing auth / permission / rate-limit failures), and the batched detail fetch no longer indexes `[0]` on a possibly-empty `resources` list. Covered by `test_raise_for_status_*` and `test_query_notifications_raises_on_api_error`.

File: external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+    def __init__(
+        self,
+        helper: OpenCTIConnectorHelper,
+        base_url: HttpUrl,
+        client_id: str,
+        client_secret: str,
+        filter_topic: str = "",
+        filter_type: str = "",
+        filter_priority: str = "",
+    ):
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py:23]: Fixed. Dropped the misleading `HttpUrl` annotation (and its import); `base_url` is now typed `str`, matching the `ConnectorSettings` value and what falconpy expects.

File: external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+        self.client_secret = client_secret
+        self.filter_topic = filter_topic
+        self.filter_type = filter_type
+        self.filter_priority = filter_priority
+
+        self.cs = CrowdstrikeRecon(
+            client_id=client_id,
+            client_secret=client_secret,
+            base_url=base_url
+        )
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py:50]: Fixed. The client secret is no longer stored as an instance attribute; it is only passed to the falconpy constructor, limiting the secret's exposure surface.

File: external-import/crowdstrike-recon/docker-compose.yml [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+      - CONNECTOR_NAME=ChangeMe
+      - CONNECTOR_SCOPE=ChangeMe
+      - CONNECTOR_LOG_LEVEL=error # optional (default: 'error')
+      - CONNECTOR_DURATION_PERIOD=ChangeMe # optional, interval in ISO-8601 format between two runs of the connector (default: 'PT1H')
+      # Custom parameters for connector
+      - CROWDSTRIKE_RECON_API_BASE_URL=ChangeMe
+      - CROWDSTRIKE_RECON_CLIENT_ID=ChangeMe
+      - CROWDSTRIKE_RECON_CLIENT_SECRET=ChangeMe
+      - CROWDSTRIKE_RECON_TLP_LEVEL=clear # optional, available values: 'clear', 'white', 'green', 'amber', 'amber+strict', 'red' (default: 'clear')
+      - CROWDSTRIKE_RECON_IMPORT_START_DATE=P10D # optional, ISO 8601 duration string specifying how far back to import data (e.g., P1D for 1 day, P7D for 7 days). Default: 'P7D'
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/docker-compose.yml]: Fixed. The docker-compose comments and value now match the real defaults: TLP `amber+strict` and import start date `P10D` (and the filter_type example uses valid item types).

File: external-import/crowdstrike-recon/src/connector/converter_to_stix.py [hunk 14]
```diff
@@ -0,0 +1,422 @@
...
+
+    def create_incident(self, notification_detail: dict) -> list:
+        """
+        :param notification_detail:
+        :return:
+        """
+        notification = notification_detail.get("notification", {})
+        stix_objects = []
+        incident_date = parse(notification.get("created_date"))
+        incident_type = "alert"
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/connector/converter_to_stix.py]: Fixed. `incident_type` is now derived from the notification `item_type` (instead of being hard-coded to "alert"), and the README description was updated to match.

File: external-import/crowdstrike-recon/src/connector/utils.py [hunk 15]
```diff
@@ -0,0 +1 @@
+#  Utilities: helper functions, classes, or modules that provide common, reusable functionality across a codebase
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/connector/utils.py:1]: Fixed. Removed the empty `utils.py` stub.

File: external-import/crowdstrike-recon/src/connector/converter_to_stix.py [hunk 16]
```diff
@@ -0,0 +1,422 @@
...
+            f"- **Category**: {reply_details.get('category') or '--'}\n"
+            f"- **Title**: {reply_details.get('title') or '--'}\n"
+            "### Reply Content\n"
+            "```\n"
+            f"{reply_details.get('content')}\n"
+            "```\n"
+        )
+        return markdown_content
+
+    def create_incident(self, notification_detail: dict) -> list:
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/connector/converter_to_stix.py:289]: Done. Added a `tests/` suite: `test_converter_to_stix.py` (each `item_type` branch plus the guarded edge cases), `test_api_client.py` (FQL build/escaping, pagination, status checks, batching), `test_settings.py`, and `test_connector.py` (run loop + work lifecycle). The CI `test` step runs them via `tests/test-requirements.txt`.

File: external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+        if type_values:
+            values = ",".join(f"'{v}'" for v in type_values)
+            fql_parts.append(f"item_type:[{values}]")
+
+        if priority_values:
+            values = ",".join(f"'{v}'" for v in priority_values)
+            fql_parts.append(f"priority:[{values}]")
+
+        # filter per updated_date
+        fql_parts.append(f"created_date:>'{from_date}'")
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py]: Fixed. Filter values are escaped with `_escape_fql_value` (single quotes are backslash-escaped) before being interpolated into the single-quoted FQL string. Covered by `test_build_fql_filter_escapes_single_quotes`.

File: external-import/crowdstrike-recon/README.md [hunk 18]
```diff
@@ -0,0 +1,249 @@
...
+
+### Requirements
+
+- Python >= 3.11
+- OpenCTI Platform >= 6.8.13
+- [`pycti`](https://pypi.org/project/pycti/) library matching your OpenCTI version
+- [`connectors-sdk`](https://github.com/OpenCTI-Platform/connectors.git@master#subdirectory=connectors-sdk) library matching your OpenCTI version
+- CrowdStrike Falcon subscription with **Recon / Exposure Monitoring** module
+- CrowdStrike API credentials (Client ID and Client Secret) with **Recon** read permissions
+
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/README.md:66]: Fixed. The README now states OpenCTI Platform >= 6.9.5, matching the manifest `support_version`.

File: external-import/crowdstrike-recon/Dockerfile [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+COPY src /opt/opencti-connector-crowdstrike-recon
+
+# Install Python modules
+# hadolint ignore=DL3003
+RUN apk update && apk upgrade && \
+    apk --no-cache add git build-base libmagic libffi-dev libxml2-dev libxslt-dev
+
+RUN cd /opt/opencti-connector-crowdstrike-recon && \
+    pip3 install --no-cache-dir -r requirements.txt && \
+    apk del git build-base
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/Dockerfile]: Fixed. The build headers (build-base, libffi-dev, libxml2-dev, libxslt-dev, git) are now installed as a `.build-deps` virtual package and removed after install, while the required runtime libraries (libffi, libxml2, libxslt, libmagic) are kept so the image stays functional but slimmer.

File: external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+
+            notification_ids.extend(resources)
+
+            total = result.get("body", {}).get("meta", {}).get("pagination", {}).get("total", 0)
+            if len(notification_ids) >= total:
+                break
+
+            offset += limit
+
+        return notification_ids
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py]: Fixed. Pagination now also breaks on a short page (`len(resources) < limit`) and is bounded by a max-pages safety guard to avoid an unbounded loop. Covered by `test_query_notifications_stops_on_short_page`.

File: external-import/crowdstrike-recon/src/requirements.txt [hunk 21]
```diff
@@ -0,0 +1,3 @@
+pycti==7.260512.0
+connectors-sdk @ git+https://github.com/OpenCTI-Platform/connectors.git@master#subdirectory=connectors-sdk
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/requirements.txt]: Fixed. Bumped `pycti` to `7.260609.0` to match `connectors-sdk` on master (and the latest release). The `connectors-sdk @ master` pin is the repository convention for connectors, so it is kept intentionally.

File: external-import/crowdstrike-recon/src/connector/connector.py [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+                notification_detail=notification_detail
+            )
+            stix_objects.extend(stix_entities)
+
+        # Ensure consistent bundle by adding the author and TLP marking
+        if stix_objects:
+            stix_objects.append(self.converter_to_stix.author)
+            stix_objects.append(self.converter_to_stix.tlp_marking)
+
+        return stix_objects, most_recent_incident_date
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/connector/connector.py]: Fixed. The method and its caller now consistently use `most_recent_alert_date`.

**Review by SamuelHassine — APPROVED** [review]
Reviewed the full new CrowdStrike Recon connector (not just the diff) and addressed the entire Copilot review pass. Hardened the converter (missing-key guards, base64 string attachment, derived incident_type), the API client (FQL quote escaping, response status checks, batched detail fetch, no stored secret), and the connector work lifecycle (is_multipart + to_processed in a finally block). Added converter / API-client / settings / connector test suites; codecov/patch and codecov/project are green, the connector test job passes, and there are 0 unresolved review threads. De-staled the branch by merging master. LGTM.

**Comment by SamuelHassine** [comment]
## Review pass - CrowdStrike Recon connector

Reviewed the full connector and addressed every Copilot thread (22, all resolved), then drove CI to green.

Highlights:

- converter_to_stix: guard highlights[] / notification / breach_summary against missing keys; default rule_name / title; base64 attachment as a string (not bytes); derive incident_type from item_type; dedicated generate_file_content.
- api_client: drop the misleading HttpUrl hint and the stored client secret; escape single quotes in FQL values; raise on non-2xx falconpy responses; pagination guard (short page + max pages); batch notification-detail lookups (removes the N+1).
- connector: fix the set-literal log meta; track the maximum alert date; open the work as is_multipart and always close it via to_processed in a finally block.
- docs/build: README support version (>= 6.9.5) and incident-type wording; docker-compose default comments; slimmer Dockerfile build deps; removed the empty utils.py; bumped pycti to 7.260609.0.
- tests: added converter, API client, settings and connector test suites.
- Merged master to de-stale the branch (it was ~1 month behind), aligning it with the current connectors-sdk and CI.

Status: all CI checks green (Test external-import/crowdstrike-recon, codecov/patch + codecov/project, Build and Commit Manifest, linters, signed commits, PR title, issue link); 0 unresolved review threads; marked ready for review.

File: external-import/crowdstrike-recon/src/connector/converter_to_stix.py [hunk 23]
```diff
@@ -0,0 +1,464 @@
...
+                definition_type="statement",
+                definition={"statement": "custom"},
+                custom_properties={
+                    "x_opencti_definition_type": "TLP",
+                    "x_opencti_definition": "TLP:AMBER+STRICT",
+                },
+            ),
+            "red": stix2.TLP_RED,
+        }
+        return mapping[level]
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/connector/converter_to_stix.py:83]: Fixed. `tlp_level="clear"` now maps to a distinct `TLP:CLEAR` MarkingDefinition (mirroring `connectors_sdk.models.tlp_marking`) instead of aliasing it to `TLP:WHITE`. Covered by `test_clear_tlp_marking_is_distinct_from_white`.

File: external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py [hunk 24]
```diff
@@ -0,0 +1,185 @@
...
+        if status_code is None or status_code >= 300:
+            errors = (result.get("body") or {}).get("errors")
+            self.helper.connector_logger.error(
+                "[API CLIENT] CrowdStrike Recon API request failed",
+                {
+                    "operation": operation,
+                    "status_code": status_code,
+                    "errors": errors,
+                },
+            )
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py:106]: Fixed. This call (and all other `connector_logger` calls in the connector) now pass the structured context via the `meta={...}` keyword instead of positionally, consistent with the rest of the repo.

File: external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py [hunk 25]
```diff
@@ -0,0 +1,185 @@
...
+        limit = 100
+        sort = "created_date|asc"
+
+        fql_filter = self._build_fql_filter(from_date=from_date)
+
+        if fql_filter:
+            self.helper.connector_logger.info(
+                "[API CLIENT] Querying notifications with FQL filter",
+                {"filter": fql_filter},
+            )
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py:130]: Fixed. Now uses `meta={...}` for the structured context.

File: external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py [hunk 26]
```diff
@@ -0,0 +1,185 @@
...
+            if len(notification_ids) >= total or len(resources) < limit:
+                break
+
+            offset += limit
+        else:
+            self.helper.connector_logger.warning(
+                "[API CLIENT] Reached pagination safety limit while querying "
+                "notifications; results may be truncated",
+                {"max_pages": self._MAX_PAGES},
+            )
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py:168]: Fixed. Now uses `meta={...}` for the structured context.

File: external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py [hunk 27]
```diff
@@ -0,0 +1,185 @@
...
+            if not resources:
+                break
+
+            notification_ids.extend(resources)
+
+            total = ((body.get("meta") or {}).get("pagination") or {}).get("total", 0)
+            # Stop once everything has been collected or the API returns a short
+            # page (fewer than ``limit`` results means there is no next page).
+            if len(notification_ids) >= total or len(resources) < limit:
+                break
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/crowdstrike_client/api_client.py]: Fixed. Pagination no longer treats a missing/zero `total` as "done": it breaks on a short page (`len(resources) < limit`) and only uses `total` when the API actually reports it (bounded by `_MAX_PAGES`). Covered by `test_query_notifications_missing_total_continues_until_short_page`.

File: external-import/crowdstrike-recon/src/connector/connector.py [hunk 28]
```diff
@@ -0,0 +1,235 @@
...
+        for notification_detail in notification_details:
+            created_date = (notification_detail.get("notification") or {}).get(
+                "created_date"
+            )
+            # Track the maximum created_date explicitly so the saved state never
+            # regresses, regardless of the order the details are returned in.
+            if created_date and (
+                most_recent_alert_date is None or created_date > most_recent_alert_date
+            ):
+                most_recent_alert_date = created_date
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/connector/connector.py]: Fixed. `_collect_intelligence` now parses `created_date` to a `datetime` for the max comparison and persists the original string, so fractional seconds / timezone offsets are compared correctly.

File: external-import/crowdstrike-recon/Dockerfile [hunk 29]
```diff
@@ -0,0 +1,20 @@
...
+
+# Copy the connector
+COPY src /opt/opencti-connector-crowdstrike-recon
+
+# Install Python modules
+# hadolint ignore=DL3003
+RUN apk update && apk upgrade && \
+    apk --no-cache add libmagic libffi libxml2 libxslt && \
+    apk --no-cache add --virtual .build-deps git build-base libffi-dev libxml2-dev libxslt-dev
+
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/Dockerfile]: Kept intentionally. `apk upgrade` patches the base image and is used by roughly half of this repository's connector Dockerfiles (77 of 152) as a security practice; the `--no-cache` flags on the `add` steps already avoid leaving an APK index cache in the image. Happy to drop it if the team prefers a stricter reproducibility stance.

File: external-import/crowdstrike-recon/config.yml.sample [hunk 30]
```diff
@@ -0,0 +1,20 @@
+opencti:
+  url: 'http://localhost'
+  token: 'ChangeMe'
+
+connector:
+  id: 'ChangeMe'
+  name: 'CrowdStrike Recon' # optional (default: 'CrowdStrike Recon')
+  scope: 'ChangeMe'
+  log_level: 'error' # optional (default: 'error')
+  duration_period: 'PT1H' #  # optional, interval in ISO-8601 format between two runs of the connector (default: 'PT1H')
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/config.yml.sample]: Fixed. Removed the duplicate `#` on the `duration_period` comment (and aligned the `filter_type` example with valid item types).

File: external-import/crowdstrike-recon/src/connector/converter_to_stix.py [hunk 31]
```diff
@@ -0,0 +1,477 @@
...
+            inc_description = self.generate_common_description(notification)
+            alert_detail = self.generate_file_content(notification_detail)
+        else:
+            self.helper.connector_logger.error(
+                "Unsupported notification type, skipping alert",
+                meta={
+                    "item_type": item_type,
+                    "notification_id": notification.get("id"),
+                },
+            )
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/connector/converter_to_stix.py]: Fixed. Unsupported notification types are now logged at `warning` level, matching the README wording.

File: external-import/crowdstrike-recon/Dockerfile [hunk 32]
```diff
@@ -0,0 +1,20 @@
...
+RUN cd /opt/opencti-connector-crowdstrike-recon && \
+    pip3 install --no-cache-dir -r requirements.txt && \
+    apk del .build-deps
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/Dockerfile]: Fixed. The `apk add` (including `.build-deps`), `pip install` and `apk del .build-deps` now run in a single RUN layer, so the build headers are actually reclaimed instead of persisting in an earlier image layer. The base-image upgrade is now `apk --no-cache upgrade`.

File: external-import/crowdstrike-recon/src/connector/settings.py [hunk 33]
```diff
@@ -0,0 +1,75 @@
...
+        description="ISO 8601 duration string specifying how far back to import alerts (e.g., P1D for 1 day, P7D for 7 days)",
+    )
+    filter_topic: str = Field(
+        description="Filter notifications by topic name(s). Comma-separated string (e.g. 'SA_BRAND,SA_THIRD_PARTY_V2'). Empty means no filtering.",
+        default="",
+    )
+    filter_type: str = Field(
+        description="Filter notifications by type(s). Comma-separated string (e.g. 'typosquatting,phishing'). Empty means no filtering.",
+        default="",
+    )
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/connector/settings.py:63]: Fixed. The `filter_type` example now uses valid item types (`typosquatting_domain,exposed_data`) in both the settings field description and the sample configs.

File: external-import/crowdstrike-recon/docker-compose.yml [hunk 34]
```diff
@@ -0,0 +1,24 @@
+version: '3'
+services:
+  connector-crowdstrike-recon:
+    image: opencti/connector-crowdstrike-recon:latest # 'template' being the name of connector's directory
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/docker-compose.yml]: Fixed. Removed the leftover "'template' being the name of connector's directory" comment from the image line.

File: external-import/crowdstrike-recon/src/connector/converter_to_stix.py [hunk 35]
```diff
@@ -0,0 +1,479 @@
...
+            return []
+
+        stix_objects = []
+        incident_date = parse(created_date)
+        # incident_type is derived from the CrowdStrike notification item_type
+        incident_type = item_type or "alert"
```
[the remark this answers is not in the document]
Comment (SamuelHassine) [reply external-import/crowdstrike-recon/src/connector/converter_to_stix.py:322]: Fixed. `create_incident` now wraps `parse(created_date)` in a try/except and logs + skips (returns []) on an unparseable timestamp, matching the guard already used in `_collect_intelligence`. Covered by `test_create_incident_unparseable_created_date_returns_empty`.

**Comment by SamuelHassine** [comment]
## Final review status

Copilot ran several follow-up passes; all are now addressed and the last pass converged with no new findings. 40 review threads were resolved in total.

Additional hardening since the first summary:

- TLP: `clear` now emits a distinct `TLP:CLEAR` marking instead of being aliased to `TLP:WHITE`.
- Logging: structured context is passed via `meta={...}` consistently.
- Pagination: a missing/zero `total` no longer truncates results (short-page + max-pages guards); empty / short / total paths are covered by tests.
- State cursor: alert dates are compared as parsed datetimes (the original string is persisted).
- Robustness: `create_incident` skips a notification with an unparseable `created_date` instead of crashing the run.
- Docker: build deps are added, used and removed in a single RUN layer; the base image is upgraded with `--no-cache`.
- Docs: README / docker-compose / config.yml.sample / settings examples are aligned, and `filter_type` uses valid item types.
- Tests: connector / converter / client / settings suites; patch coverage is ~98% on the connector modules.

Status: all CI checks green (tests, codecov/patch + codecov/project, Build and Commit Manifest, linters, signed commits, PR title, issue link); 0 unresolved review threads; approved; marked ready for review.

Note: GitHub still reports `mergeStateStatus: BLOCKED` even though the PR is approved, fully green and has no unresolved threads. This looks like the branch ruleset (the required Copilot code review registering on the exact head commit) rather than anything in the diff; it should clear shortly or on the maintainer's merge action.

File: external-import/crowdstrike-recon/__metadata__/connector_manifest.json [hunk 36]
```diff
@@ -0,0 +1,22 @@
...
+    "Third-Party & Supply Chain Oversight"
+  ],
+  "verified": false,
+  "last_verified_date": null,
+  "playbook_supported": false,
+  "max_confidence_level": 50,
+  "support_version": ">=6.9.5",
+  "subscription_link": null,
+  "source_code": "https://github.com/OpenCTI-Platform/connectors/tree/master/external-import/crowdstrike-recon",
+  "manager_supported": false,
```
Comment (ncarenton) [inline external-import/crowdstrike-recon/__metadata__/connector_manifest.json]: @romain-filigran shouldn't we support catalog deployment ?

Comment (ncarenton) [reply external-import/crowdstrike-recon/__metadata__/connector_manifest.json]: Fixed

**Comment by ncarenton** [comment]
Tested catalog deployment locally, had to add default scope. Cc @romain-filigran 

image

**Review by ncarenton — APPROVED** [review]
